# EDA and Data Cleaning

This dataset is about the Water Wells in Tanzania. This data will need to be prepared and cleaned as it will be used in a classifer to predict the condition of a water well. The data can be found [here](https://www.drivendata.org/competitions/7/pump-it-up-data-mining-the-water-table/data/)
<hr style="border:2px solid magenta">

## Table of Contents: <a id ="title"></a>
- [Imports](#imports)
- [Opening Data](#opening)
- [Helper Functions](#functions)
- [Cleaning](#cleaning)
- [Exports](#exports)

### [Imports](#title) <a id = 'imports'></a>
<hr style="border:2px solid magenta">

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import datetime as dt
%matplotlib inline

### [Opening Data](#title) <a id = 'opening'></a>
<hr style="border:2px solid magenta"> 
TrainingValues.csv are the values of the waterpumps. I used parse_dates to convert that column into a datetime which will make  
data wrangling easier later on. 

In [2]:
x = pd.read_csv('./Data/TrainingValues.csv', parse_dates=['date_recorded'])
y = pd.read_csv('./Data/TrainingLabels.csv')

Taking a look at x

In [3]:
x.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 59400 entries, 0 to 59399
Data columns (total 40 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   id                     59400 non-null  int64         
 1   amount_tsh             59400 non-null  float64       
 2   date_recorded          59400 non-null  datetime64[ns]
 3   funder                 55765 non-null  object        
 4   gps_height             59400 non-null  int64         
 5   installer              55745 non-null  object        
 6   longitude              59400 non-null  float64       
 7   latitude               59400 non-null  float64       
 8   wpt_name               59400 non-null  object        
 9   num_private            59400 non-null  int64         
 10  basin                  59400 non-null  object        
 11  subvillage             59029 non-null  object        
 12  region                 59400 non-null  object        
 13  r

In [4]:
x.describe()

id     amount_tsh    gps_height     longitude      latitude  \
count  59400.000000   59400.000000  59400.000000  59400.000000  5.940000e+04   
mean   37115.131768     317.650385    668.297239     34.077427 -5.706033e+00   
std    21453.128371    2997.574558    693.116350      6.567432  2.946019e+00   
min        0.000000       0.000000    -90.000000      0.000000 -1.164944e+01   
25%    18519.750000       0.000000      0.000000     33.090347 -8.540621e+00   
50%    37061.500000       0.000000    369.000000     34.908743 -5.021597e+00   
75%    55656.500000      20.000000   1319.250000     37.178387 -3.326156e+00   
max    74247.000000  350000.000000   2770.000000     40.345193 -2.000000e-08   

        num_private   region_code  district_code    population  \
count  59400.000000  59400.000000   59400.000000  59400.000000   
mean       0.474141     15.297003       5.629747    179.909983   
std       12.236230     17.587406       9.633649    471.482176   
min        0.000000      1.000000       0.000000      0.000000   
25%        0.000000      5.000000       2.000000      0.000000   
50%        0.000000     12.000000       3.000000     25.000000   
75%        0.000000     17.000000       5.000000    215.000000   
max     1776.000000     99.000000      80.000000  30500.000000   

       construction_year  
count       59400.000000  
mean         1300.652475  
std           951.620547  
min             0.000000  
25%             0.000000  
50%          1986.000000  
75%          2004.000000  
max          2013.000000

In [5]:
x.columns

Index(['id', 'amount_tsh', 'date_recorded', 'funder', 'gps_height',
       'installer', 'longitude', 'latitude', 'wpt_name', 'num_private',
       'basin', 'subvillage', 'region', 'region_code', 'district_code', 'lga',
       'ward', 'population', 'public_meeting', 'recorded_by',
       'scheme_management', 'scheme_name', 'permit', 'construction_year',
       'extraction_type', 'extraction_type_group', 'extraction_type_class',
       'management', 'management_group', 'payment', 'payment_type',
       'water_quality', 'quality_group', 'quantity', 'quantity_group',
       'source', 'source_type', 'source_class', 'waterpoint_type',
       'waterpoint_type_group'],
      dtype='object')

Seeing how much nulls is in x

In [6]:
percent_missing = x.isnull().sum()*100 /len(x)
missing_value_df = pd.DataFrame({'column_name' : x.columns,
                                 'percent_missing': percent_missing})
missing_value_df

column_name  percent_missing
id                                        id         0.000000
amount_tsh                        amount_tsh         0.000000
date_recorded                  date_recorded         0.000000
funder                                funder         6.119529
gps_height                        gps_height         0.000000
installer                          installer         6.153199
longitude                          longitude         0.000000
latitude                            latitude         0.000000
wpt_name                            wpt_name         0.000000
num_private                      num_private         0.000000
basin                                  basin         0.000000
subvillage                        subvillage         0.624579
region                                region         0.000000
region_code                      region_code         0.000000
district_code                  district_code         0.000000
lga                                      lga         0.000000
ward                                    ward         0.000000
population                        population         0.000000
public_meeting                public_meeting         5.612795
recorded_by                      recorded_by         0.000000
scheme_management          scheme_management         6.526936
scheme_name                      scheme_name        47.417508
permit                                permit         5.144781
construction_year          construction_year         0.000000
extraction_type              extraction_type         0.000000
extraction_type_group  extraction_type_group         0.000000
extraction_type_class  extraction_type_class         0.000000
management                        management         0.000000
management_group            management_group         0.000000
payment                              payment         0.000000
payment_type                    payment_type         0.000000
water_quality                  water_quality         0.000000
quality_group                  quality_group         0.000000
quantity                            quantity         0.000000
quantity_group                quantity_group         0.000000
source                                source         0.000000
source_type                      source_type         0.000000
source_class                    source_class         0.000000
waterpoint_type              waterpoint_type         0.000000
waterpoint_type_group  waterpoint_type_group         0.000000

In [7]:
x.date_recorded.describe(datetime_is_numeric=True)

count                            59400
mean     2012-03-29 09:11:33.818181888
min                2002-10-14 00:00:00
25%                2011-04-01 00:00:00
50%                2012-10-10 00:00:00
75%                2013-02-09 00:00:00
max                2013-12-03 00:00:00
Name: date_recorded, dtype: object

### [Helper Functions](#title) <a id = 'functions'></a>
<hr style="border:2px solid magenta">

<span style = 'color:green; font-size:13pt'>selecting_top</span>  
**parameters**  
><span style = 'color:blue'>df</span> &rarr; dataframe  
><span style = 'color:blue'>column</span> &rarr; the name of a column in the dataframe  
><span style = 'color:blue'>num</span> &rarr; a number threshold

***

**returns** 
a dataframe  

***
**how it works**  
* performs a value counts of the column of the dataframe
* uses a lambda function to select the values that are occur over the number
* converts this to a dataframe and grab the index values which are the names of the columns

In [8]:
def selecting_top(df, column, num):
    return df[column].value_counts()[lambda x : x > num].to_frame().index.values

<hr style="border:2px solid orange">

<span style = 'color:green; font-size:13pt'>apply_top</span>  
**parameters**  
><span style = 'color:blue'>df</span> &rarr; dataframe  
><span style = 'color:blue'>column</span> &rarr; the name of a column in the dataframe  
><span style = 'color:blue'>names</span> &rarr; the names of values to keep  

***

**returns** a dataframe

***
**how it works** 
* performs apply function on the column of the dataframe
* uses a lambda function to update the row value in the colum. If the value appears in the list names, keep the same name
if not replace it with 'Other'
* converts this to a dataframe and grab the index values which are the names of the columns

In [9]:
def apply_top(df, column, names):
    df[column] = df[column].apply(lambda x: x if x in names else 'Other')

<hr style="border:2px solid orange">

<span style = 'color:green; font-size:13pt'>select_apply</span>  
**parameters**  
><span style = 'color:blue'>df</span> &rarr; dataframe  
<span style = 'color:blue'>column</span> &rarr; the name of a column in the dataframe  
<span style = 'color:blue'>num</span> &rarr; number threshold  

***

**returns** a dataframe
***
**how it works** 
* creates a variable named values equal to the output of the selecting_top function
* uses apply_top function with this values variable

In [10]:
def select_apply(df, column, num):
    values = selecting_top(df, column, num)
    apply_top(df,column, values)

<hr style="border:2px solid orange">

<span style = 'color:green; font-size:13pt'>filling_na</span>  
**parameters**  
><span style = 'color:blue'>df</span> &rarr; dataframe  
<span style = 'color:blue'>column</span> &rarr; the name of a column in the dataframe  

***

**returns** a dataframe

***
**how it works** 
* uses fillna to replace missing values with 'Unknown'

In [11]:
def filling_na(df, column):
    df[column] = df[column].fillna('Unknown')

<hr style="border:2px solid orange">

<span style = 'color:green; font-size:13pt'>day_update</span>  
**parameters**  
><span style = 'color:blue'>df</span> &rarr; dataframe  

***

**returns** a dataframe

***
**how it works** 
* finds the time delta between the date in df and 12/3/2013
* converts it into days then to an integer

In [12]:
def day_update(df):
    df['days_since_recorded'] = (dt.datetime(2013,12,3) - (df.date_recorded)).dt.days.astype(int)
    df.drop(['date_recorded'], axis = 1, inplace=True)

<hr style="border:2px solid orange">

<span style = 'color:green; font-size:13pt'>year_update</span>  
**parameters**  
><span style = 'color:blue'>df</span> &rarr; dataframe  

***

**returns** a dataframe

***
**how it works** 
* if statement to verify that the value is a string
* create bins for the decades
* create a list of labels
* use pd.cut to transform it

In [13]:
def year_update(df):
    if not isinstance(df.construction_year[2],str):
        bins = [-1, 1960, 1970,1980,1990,2000,2010,2020]
        labels = ['Unknown','60s', '70s', '80s', '90s', '00s', '10s']
        x['construction_year'] = pd.cut(df.construction_year, 
                                    bins = bins, 
                                    labels = labels, 
                                    right = False)

### [Cleaning Process](#title) <a id = 'cleaning'></a>
<hr style="border:2px solid magenta">

#### Creating variable drop that keep tracks of the columns we plan to drop

In [14]:
maybe_drop = []
confirmed_drop = []

<hr style="border:2px solid orange">
Updating days so we can see how "old" the well is

In [15]:
day_update(x)
x['days_since_recorded'].describe()

count    59400.000000
mean       613.616970
std        334.216374
min          0.000000
25%        297.000000
50%        419.000000
75%        977.000000
max       4068.000000
Name: days_since_recorded, dtype: float64

<hr style="border:2px solid orange">

Using select_apply to update the x with names of funders that appear more than 300 times

In [16]:
select_apply(x, 'funder', 300)

<hr style="border:2px solid orange">

Looking at gps_height column  


In [17]:
x.gps_height.value_counts()

 0       20438
-15         60
-16         55
-13         55
-20         52
         ...  
 2285        1
 2424        1
 2552        1
 2413        1
 2385        1
Name: gps_height, Length: 2428, dtype: int64

In [18]:
maybe_drop.append('gps_height')

I dropped this column, but they might be useful to look at a later time

<hr style="border:2px solid orange">


Looking at installer column

In [19]:
x.installer.value_counts()

DWE                    17402
Government              1825
RWE                     1206
Commu                   1060
DANIDA                  1050
                       ...  
Mama Hamisa                1
Pet corporation Ltd        1
Maerere                    1
Mr Chi                     1
Malec                      1
Name: installer, Length: 2145, dtype: int64

In [20]:
select_apply(x, 'installer', 200)

Using select_apply to update the x with names of installers that appear more than 200 times

<hr style="border:2px solid orange">



Looking at latitude,longitude, wpt_name, num_private, region_code, district_code

In [21]:
x.latitude.value_counts().head()

-2.000000e-08    1812
-6.985842e+00       2
-3.797579e+00       2
-6.981884e+00       2
-7.104625e+00       2
Name: latitude, dtype: int64

In [22]:
x.longitude.value_counts().head()

0.000000     1812
37.540901       2
33.010510       2
39.093484       2
32.972719       2
Name: longitude, dtype: int64

In [23]:
x.wpt_name.value_counts().head()

none         3563
Shuleni      1748
Zahanati      830
Msikitini     535
Kanisani      323
Name: wpt_name, dtype: int64

In [24]:
x.num_private.value_counts().head()

0    58643
6       81
1       73
5       46
8       46
Name: num_private, dtype: int64

In [25]:
x.region_code.value_counts().head()

11    5300
17    5011
12    4639
3     4379
5     4040
Name: region_code, dtype: int64

In [26]:
x.district_code.value_counts().head()

1    12203
2    11173
3     9998
4     8999
5     4356
Name: district_code, dtype: int64

In [27]:
maybe_drop.extend(['latitude', 'longitude', 'wpt_name','region_code', 'district_code'])
confirmed_drop.extend(['num_private'])

These columns are associated with locations of the well. I dropped these but they might be useful to look at a later time.  
**NOTE** num_private has no valuable information upon further inspection on board/discussion post

<hr style="border:2px solid orange">

Looking at the basin column

In [28]:
x.basin.value_counts(dropna=False)

Lake Victoria              10248
Pangani                     8940
Rufiji                      7976
Internal                    7785
Lake Tanganyika             6432
Wami / Ruvu                 5987
Lake Nyasa                  5085
Ruvuma / Southern Coast     4493
Lake Rukwa                  2454
Name: basin, dtype: int64

This looks good. No need to use select_top on this column
<hr style="border:2px solid orange">
Now let's take a look at the subvillages

In [29]:
x.subvillage.value_counts(dropna=False)

Madukani        508
Shuleni         506
Majengo         502
Kati            373
NaN             371
               ... 
Ishongo Kati      1
Bukula            1
Ipuli B           1
Majevu B          1
Kwa Bonda         1
Name: subvillage, Length: 19288, dtype: int64

In [30]:
filling_na(x, 'subvillage')

I noted that were some missing values so I looked at the value counts with missing values. I decided to fill these with Unknown because it might affect Madukani predictions. 
<hr style="border:2px solid orange">

Looking at region columns

In [31]:
x.region.value_counts(dropna=False)

Iringa           5294
Shinyanga        4982
Mbeya            4639
Kilimanjaro      4379
Morogoro         4006
Arusha           3350
Kagera           3316
Mwanza           3102
Kigoma           2816
Ruvuma           2640
Pwani            2635
Tanga            2547
Dodoma           2201
Singida          2093
Mara             1969
Tabora           1959
Rukwa            1808
Mtwara           1730
Manyara          1583
Lindi            1546
Dar es Salaam     805
Name: region, dtype: int64

This looks good so I left it alone
<hr style="border:2px solid orange">

Looking at ward

In [32]:
x.ward.value_counts(dropna=False)

Igosi        307
Imalinyi     252
Siha Kati    232
Mdandu       231
Nduruma      217
            ... 
Mawenzi        1
Themi          1
Nsemulwa       1
Rasbura        1
Ifinga         1
Name: ward, Length: 2092, dtype: int64

I decided to leave this alone. However using select_apply function might be useful in for column. 
<hr style="border:2px solid orange">

Taking a look at public_meeting

In [33]:
x.public_meeting.value_counts(dropna=False)

True     51011
False     5055
NaN       3334
Name: public_meeting, dtype: int64

In [34]:
filling_na(x, 'public_meeting')

In [35]:
maybe_drop.extend(['public_meeting'])

I do not really know what this column means completely. I decided to use filling_na.  
If I have to time, would love to see the effect of this column on the performance of some of the models by dropping it.
<hr style="border:2px solid orange">

Taking a look at scheme_management

In [36]:
x.scheme_management.value_counts(dropna=False)

VWC                 36793
WUG                  5206
NaN                  3877
Water authority      3153
WUA                  2883
Water Board          2748
Parastatal           1680
Private operator     1063
Company              1061
Other                 766
SWC                    97
Trust                  72
None                    1
Name: scheme_management, dtype: int64

In [37]:
select_apply(x, 'scheme_management', 100)

Used select_apply with 100 for some reason. **currently writing this comment afterwards**  
Must be a typo. Should be 1000
<hr style="border:2px solid orange">

Looking at permit column

In [38]:
x.permit.value_counts(dropna=False)

True     38852
False    17492
NaN       3056
Name: permit, dtype: int64

In [39]:
filling_na(x, 'permit')

Filling_na for this column
<hr style="border:2px solid orange">

Looking at the scheme

In [40]:
x.scheme_name.value_counts()

K                            682
None                         644
Borehole                     546
Chalinze wate                405
M                            400
                            ... 
Kwemkingu                      1
Mradi wa maji wa matalawe      1
BL Moniko Shule                1
Rain Water Harvesting          1
Tove -mtwango                  1
Name: scheme_name, Length: 2696, dtype: int64

In [41]:
confirmed_drop.extend(['scheme_name'])

Does not seem like useful information as None has 644 counts and undiscernable names like K and M. dropped
<hr style="border:2px solid orange">

looking at construction year

In [42]:
x.construction_year.value_counts().head()

0       20709
2010     2645
2008     2613
2009     2533
2000     2091
Name: construction_year, dtype: int64

In [43]:
year_update(x)
x.construction_year.value_counts()

Unknown    20709
00s        15330
90s         7678
80s         5578
10s         5161
70s         4406
60s          538
Name: construction_year, dtype: int64

Decided to bin this by decades. Kept the unknown because it might be existing wells before 60s as well as some that were not recorded. 
<hr style="border:2px solid orange">

Lets look at the extractions

In [44]:
x.extraction_type.value_counts().head()

gravity        26780
nira/tanira     8154
other           6430
submersible     4764
swn 80          3670
Name: extraction_type, dtype: int64

In [45]:
x.extraction_type_group.value_counts().head()

gravity        26780
nira/tanira     8154
other           6430
submersible     6179
swn 80          3670
Name: extraction_type_group, dtype: int64

In [46]:
x.extraction_type_class.value_counts().head()

gravity        26780
handpump       16456
other           6430
submersible     6179
motorpump       2987
Name: extraction_type_class, dtype: int64

In [47]:
confirmed_drop.extend(['extraction_type_group', 'extraction_type_class'])

These are duplicates so I am going to drop two of these.
<hr style="border:2px solid orange">

Lets take a look at managements

In [48]:
x.management.value_counts().head()

vwc                 40507
wug                  6515
water board          2933
wua                  2535
private operator     1971
Name: management, dtype: int64

In [49]:
x.management_group.value_counts().head()

user-group    52490
commercial     3638
parastatal     1768
other           943
unknown         561
Name: management_group, dtype: int64

These look good!
<hr style="border:2px solid orange">

Look at payments

In [50]:
x.payment_type.value_counts().head(10)

never pay     25348
per bucket     8985
monthly        8300
unknown        8157
on failure     3914
annually       3642
other          1054
Name: payment_type, dtype: int64

In [51]:
x.payment.value_counts().head(10)

never pay                25348
pay per bucket            8985
pay monthly               8300
unknown                   8157
pay when scheme fails     3914
pay annually              3642
other                     1054
Name: payment, dtype: int64

In [52]:
confirmed_drop.extend(['payment_type'])

These are duplicates so I double it and pass it to the next guy.
<hr style="border:2px solid orange">

Lets look at quality and quantity

In [53]:
x.quality_group.value_counts().head(10)

good        50818
salty        5195
unknown      1876
milky         804
colored       490
fluoride      217
Name: quality_group, dtype: int64

In [54]:
x.water_quality.value_counts().head(10)

soft                  50818
salty                  4856
unknown                1876
milky                   804
coloured                490
salty abandoned         339
fluoride                200
fluoride abandoned       17
Name: water_quality, dtype: int64

In [55]:
x.quantity.value_counts().head()

enough          33186
insufficient    15129
dry              6246
seasonal         4050
unknown           789
Name: quantity, dtype: int64

In [56]:
x.quantity_group.value_counts().head()

enough          33186
insufficient    15129
dry              6246
seasonal         4050
unknown           789
Name: quantity_group, dtype: int64

In [57]:
confirmed_drop.extend(['quality_group', 'quantity_group'])
maybe_drop.extend(['quantity'])

Water quality and quality are very similar but water quality provides more information. I dropped the quality.  
Quantity and quantity group are duplicates so I dropped one

<hr style="border:2px solid orange">

Look at source

In [58]:
x.source.value_counts().head(10)

spring                  17021
shallow well            16824
machine dbh             11075
river                    9612
rainwater harvesting     2295
hand dtw                  874
lake                      765
dam                       656
other                     212
unknown                    66
Name: source, dtype: int64

In [59]:
x.source_type.value_counts().head(10)

spring                  17021
shallow well            16824
borehole                11949
river/lake              10377
rainwater harvesting     2295
dam                       656
other                     278
Name: source_type, dtype: int64

In [60]:
x.source_class.value_counts().head(10)

groundwater    45794
surface        13328
unknown          278
Name: source_class, dtype: int64

In [61]:
confirmed_drop.extend(['source_type', 'source_class'])

Seems granular. Keep the one with the most descriptive values which is source
<hr style="border:2px solid orange">

Lets look at water point

In [62]:
x.waterpoint_type.value_counts().head(10)

communal standpipe             28522
hand pump                      17488
other                           6380
communal standpipe multiple     6103
improved spring                  784
cattle trough                    116
dam                                7
Name: waterpoint_type, dtype: int64

In [63]:
x.waterpoint_type_group.value_counts().head(10)

communal standpipe    34625
hand pump             17488
other                  6380
improved spring         784
cattle trough           116
dam                       7
Name: waterpoint_type_group, dtype: int64

In [64]:
confirmed_drop.extend(['waterpoint_type_group'])

Seems granular. Keep the one with the most descriptive values which is waterpoint_type
<hr style="border:2px solid orange">

Lets look at recorded_by

In [65]:
x.recorded_by.value_counts()

GeoData Consultants Ltd    59400
Name: recorded_by, dtype: int64

In [66]:
confirmed_drop.extend(['recorded_by'])

I dropped this because its the same value
<hr style="border:2px solid orange">

Lets look at the the drop variables

In [67]:
confirmed_drop

['num_private',
 'scheme_name',
 'extraction_type_group',
 'extraction_type_class',
 'payment_type',
 'quality_group',
 'quantity_group',
 'source_type',
 'source_class',
 'waterpoint_type_group',
 'recorded_by']

In [68]:
maybe_drop

['gps_height',
 'latitude',
 'longitude',
 'wpt_name',
 'region_code',
 'district_code',
 'public_meeting',
 'quantity']

In [69]:
confirmed_drop.extend(['id'])

Adding id to this because it is a unique identifier

### [Exporting dataframes to CSVs](#title) <a id = 'exports'></a>
<hr style="border:2px solid magenta">

Merge x and y to make exporting and importing easier.  
Decided to make two versions one with confirmed drops and one with both confirmed and maybe drops.  
If I have time, I would do an iterative model on both

In [70]:
train = pd.merge(x, y, on='id')

Looking to see if we have any null values after cleaning

In [71]:
train.drop(confirmed_drop, axis =1, inplace=True)

In [72]:
train.to_csv('./Data/TrainCleaned1.csv', index = False)

In [73]:
train.drop(maybe_drop, axis=1, inplace=True)
train.to_csv('./Data/TrainCleaned2.csv', index = False)

TrainCleaned1.csv is only with confirmed drop and the TrainCleaned2.csv is with both.

<hr style="border:2px solid orange">
Now do it with the test values

In [74]:
testv = pd.read_csv('./Data/TestValues.csv',parse_dates=['date_recorded'])
testlabel = pd.read_csv('./Data/TestPredict.csv')
test = pd.merge(testv, testlabel, on ='id')
test.drop(confirmed_drop,axis = 1, inplace=True)
day_update(test)
select_apply(test, 'funder', 300)
select_apply(test, 'scheme_management', 100)
select_apply(test, 'installer', 200)
year_update(test)
filling_na(test, 'public_meeting')
filling_na(test, 'permit')
filling_na(test, 'subvillage')

test.to_csv('./Data/TestValuesCleaned1.csv', index = False)
test.drop(maybe_drop,axis = 1, inplace=True)

test.to_csv('./Data/TestValuesCleaned2.csv', index = False)

TestCleaned1.csv is only with confirmed drop and the TestCleaned2.csv is with both.

We model using these datasets in [Modeling-v2](https://github.com/irwin-lam/PumpItUp/blob/main/Modeling-v2.ipynb)

<hr style="border:2px solid orange">
Some sanity checks

In [75]:
train.shape

(59400, 21)

In [76]:
test.shape

(14850, 21)